Version 2.1.6!  New custom serializer.

larryhastings opened the pull request:
Bugfix: the hashed_ballots_tiebreaker assumed
that marshal.dumps produced an identical bytes
string given identical inputs.  This is not true!
We also apparently can't rely on pickle.dumps
to be deterministic.  So, starvote now has
its own bespoke--and completely deterministic--simple binary serializer, called starvote_custom_serializer. It's tailor-made for the needs of starvote and
isn't useful for anybody else.  But it does guarantee that hashed_ballots_tiebreaker will now produce
identical results across all supported Python
versions, across all architectures, regardless of
optimization level.

Thanks to Petr Viktorin for the bug report!  Fixes #10.

larryhastings committed:
Version 2.1.6!  New custom serializer.

Bugfix: the hashed_ballots_tiebreaker assumed
that marshal.dumps produced an identical bytes
string given identical inputs.  This is not true!
We also apparently can't rely on pickle.dumps
to be deterministic.  So, starvote now has
its own bespoke--and completely deterministic--simple
binary serializer, called starvote_custom_serializer.
It's tailor-made for the needs of starvote and
isn't useful for anybody else.  But it does guarantee
that hashed_ballots_tiebreaker will now produce
identical results across all supported Python
versions, across all architectures, regardless of
optimization level.

Thanks to Petr Viktorin for the bug report!  Fixes #10.

encukou commented on starvote/__init__.py:
```diff
@@ -475,7 +574,7 @@ class hashed_ballots_tiebreaker(Tiebreaker):
     """
     def __init__(self, *,
             counter=1, hash='sha3_512', Random=random.Random,
-            serializer=marshal.dumps, shuffles=3,
```
You can remove the other occurences of `marshal` in this file: the import and the docstring note.
There's also a mention in the README that'll now be outdated.

larryhastings replied:
Done

encukou commented on starvote/__init__.py:
```diff
@@ -430,6 +430,105 @@ def __call__(self, options, tie, desired, exception):
+    assert isinstance(ballots[0], list)
+    assert isinstance(ballots[0][0], tuple)
+    assert len(ballots[0][0]) == 2
+    assert isinstance(ballots[0][0][0], str)
```
The README says, at one point, *The candidate can be any hashable Python value*.
Now it is restricted to `str` without control characters. That's a weird implementation detail leaking into the API. Could they be just strings? You could drop the limitation: this format doesn't need to be *parseable* nor 100% *collision-resistant*.

larryhastings replied:
I like the cut of your pedantic jib!  I now quote control characters by adding an `\x1a` in front.  My faux serialization format remains parseable.

encukou reviewed (commented):
LGTM, but it needs a few documentation changes, and perhaps dropping a limitation.

larryhastings committed:
Fixes, based on a review from Petr Viktorin.

Thanks, Petr!

larryhastings committed:
Added starvote_custom_deserializer.

Added starvote_custom_deserializer, which allows us to
test that starvote_custom_serializer is working correctly.
(Not that it matters--it just needs to produce *something*
that's approximately unique-ish for these ballots.  But now
we know for certain!)

larryhastings commented:
I went ahead and wrote a custom deserializer to match my custom serializer.  Now we can be reasonably certain that the custom serializer is working as intended, thank goodness!

larryhastings committed:
Fix coverage!  Back to 100%.

encukou commented:
My pedantic jib¹⁾ remains, as this bit of README is out of date:

> The candidate can be any hashable Python value;

The serializer demands strings.


¹⁾ *(a word I [looked up](https://www.merriam-webster.com/dictionary/jib#dictionary-entry-2) but still don't quite get)*

larryhastings committed:
Warn users that hashed_ballots requires str & int.

Raise exceptions if they aren't, and test those exceptions too.
Holding steady at 100% coverage.

larryhastings commented:
"the cut of your jib" is an old-timey expression meaning "your general style / demeanor".  The only way you ever hear it is the sentence "I like the cut of your jib!", which means "I like your style!"  (I don't understand the literal phrase, it sounds like you're saying you like how they styled one particular sail on their ship.)

> The serializer demands strings.

Good point!  I'll mention that.  Hopefully that's a restriction folks can live with.

I believe the rest of the API is still sufficiently agnostic that any hashable value will work.  You're only restricted to candidates being strings if you use the `hashable_ballots_tiebreaker`.

larryhastings committed:
README mentions hashed_ballot type restrictions.

encukou commented:
Ah! Silly me, looking up individual words in an English phrase. I should know better than to do that :)

LGTM.

larryhastings commented:
Thanks, Petr!  I appreciate all your smart feedback.
